### DISCUSSION TITLE
What nodes are allowed to be attribute values?

### DISCUSSION THREAD

**Discussion by slanden** [pdml-lang/pdml-lang.github.io#14] [Q&A] [answered]
I saw one example of something like
```
[abc (
  path = [u:get example_path]
)]
```
but I didn't come across any other details about this.

**Comment by pdml-lang** [comment, marked as answer]
All [Extension Nodes](https://pdml-lang.dev/docs/extensions/user_manual/index.html#extension_nodes) are supported in quoted and unquoted attribute values.

Note: I will soon (probably this month) publish a new PDML version, including a new chapter that specifies PDML attributes.

**Reply by slanden** [reply]
Only Extension nodes? If so, how come?

I mean, I wouldn't want to nest an entire node tree under an attribute, but some nodes might make sense. Maybe nodes that don't have grandchild nodes.

Then again, what if the attribute "expects" a non-primitive, like an object or an array. I'm thinking of how many front end frameworks use attributes to pass data to components.

**Edit**
This is a little off topic, but perhaps your attributes write-up could also mention their usage with extension nodes. If they're allowed in them, I think there's a potential conflict in script nodes. How does PDML know the attributes section, `()`, is not part of the script's source code?

**Reply by pdml-lang** [reply]
> Only Extension nodes? If so, how come?
> I mean, I wouldn't want to nest an entire node tree under an attribute, but some nodes might make sense.

You _can_ store structured data (including PDML nodes) in attributes, but then you must explicitly parse the value.
But you probably _shouldn't_ do this, unless there is a good reason for it.

Explanation:

PDML does not specify how to use attributes.

As a rule of thumb, attributes are typically used only to store simple meta-data (like in HTML), e.g. scalar values like strings, numbers, booleans, enumerateds, etc.

However, noting prevents you to store structured data (of any complexity) in attributes, if you have a good reason to do so.

For example:
```
[image dimensions="[width 200][height 100][units px]" border_color="rgb(0,0,255)"]
```

The PDML parser just parses a string for each attribute value.
You can then explicitly convert this string into a value of any type required in your application.
For example, you _could_ store a JSON/PDML/XML document in an attribute.
But then you need to parse the string value (returned by the PDML parser) into a typed value (e.g. an XML tree), using a dedicated parser in your application.

In a nutshell:
- Attributes are typically used to store simple meta-data.
- Nodes should preferably be used to store structured data.
- In rare situations, attributes can be used to store structured data. The PDML parser just parses a string which can then be parsed into a typed value by the client application.

Note: As said already, I will soon publish a new PDML version with a dedicated chapter for attributes.
